ccm-connector/src/main/java/com/sequenceiq/cloudbreak/ccmimpl/cloudinit/DefaultCcmV2ParameterSupplier.java

```diff
@@ -26,24 +28,27 @@
     @Inject
     private CcmV2ManagementClient ccmV2Client;
 
-    public CcmV2Parameters getCcmV2Parameters(@Nonnull String accountId, @Nonnull String clusterGatewayDomain, @Nonnull String agentKeyId) {
+    public CcmV2Parameters getCcmV2Parameters(@Nonnull String accountId, @Nonnull Optional<String> environmentCrnOpt,
+        @Nonnull String clusterGatewayDomain, @Nonnull String agentKeyId) {
         String requestId = MDCBuilder.getOrGenerateRequestId();
 
         InvertingProxy invertingProxy = ccmV2Client.awaitReadyInvertingProxyForAccount(requestId, accountId);
-        InvertingProxyAgent invertingProxyAgent = ccmV2Client.registerInvertingProxyAgent(requestId, accountId, clusterGatewayDomain, agentKeyId);
+        InvertingProxyAgent invertingProxyAgent = ccmV2Client.registerInvertingProxyAgent(requestId, accountId, environmentCrnOpt,
+                clusterGatewayDomain, agentKeyId);
         validateCcmV2ConfigResponse(invertingProxy, invertingProxyAgent);
 
-        LOGGER.debug("CcmV2Config successfully retrieved InvertingProxyHost : '{}', InvertingProxyStatus : '{}' InvertingProxyAgentCrn '{}'",
-                invertingProxy.getHostname(), invertingProxy.getStatus(), invertingProxyAgent.getAgentCrn());
+        LOGGER.debug("CcmV2Config successfully retrieved InvertingProxyHost: '{}', InvertingProxyStatus: '{}', InvertingProxyAgentCrn: '{}', " +
+                        "EnvironmentCrnOpt: '{}'", invertingProxy.getHostname(), invertingProxy.getStatus(),
+                invertingProxyAgent.getAgentCrn(), Optional.of(invertingProxyAgent.getEnvironmentCrn()));
 
-        return new DefaultCcmV2Parameters(invertingProxy.getHostname(), invertingProxy.getCertificate(),
-                invertingProxyAgent.getAgentCrn(), agentKeyId,
+        return new DefaultCcmV2Parameters(invertingProxy.getHostname(), invertingProxy.getCertificate(), invertingProxyAgent.getAgentCrn(), agentKeyId,
                 invertingProxyAgent.getEncipheredPrivateKey(), invertingProxyAgent.getCertificate());
     }
 
     private void validateCcmV2ConfigResponse(InvertingProxy invertingProxy, InvertingProxyAgent invertingProxyAgent) {
         checkArgument(StringUtils.isNotEmpty(invertingProxy.getHostname()), "InvertingProxy Hostname is not initialized.");
         checkArgument(StringUtils.isNotEmpty(invertingProxy.getCertificate()), "InvertingProxy Certificate is not initialized.");
+        // environmentCrn is an optional parameter and would be available only for a freeipa registration; and hence, should not be checked.
```

Comment:
Is this comment misplaced? I dont see `environmentCrn` in the below line.

Reply:
I agree, I don't think we need this comment

Reply:
anyway, as it's optional it still must not be null as it would cause nullpointer later

Reply:
The check is happening over the protobuf object, which is null if not returned from the CCM service.

Reply:
The check is happening over the protobuf object, which is null if not returned from the CCM service. I would drop the comment though.